src/scarlet/license_finder/report.clj

```diff
@@ -3,25 +3,65 @@
             [clojure.data.csv :as csv]
             [clojure.java.io :as io]))
 
-(defn- write-csv [file-path deps]
-  (with-open [writer (io/writer (fs/file file-path))]
-    (->> deps
-         (map (fn [{:keys [version type license]
-                    dep-name :name}]
-                [dep-name
-                 version
-                 (name type)
-                 (:id license)
-                 (:name license)
-                 (:url license)]))
-         (cons ["Dependency" "Version" "Type"
-                "License ID" "License Name" "License URL"])
-         (csv/write-csv writer))))
-
-(defn write-report [{:keys [project out-file]} deps]
-  (let [out-file (or out-file
-                     (fs/normalize (fs/path "./target/licenses" (str project ".csv"))))]
-    (when-let [dir (fs/parent out-file)]
+(def headers ["Dependency" "Version" "Type" "License ID" "License Name" "License URL"])
+
+(defn- dep->vec
+  "Return a flat vector with dependency's data out of a nested map"
+  [{:keys [version type license] dep-name :name}]
+  [dep-name
+   version
+   (name type)
+   (:id license)
+   (:name license)
+   (:url license)])
+
+(defn- prep-path
+  [{:keys [path]}]
+  (let [path (-> path
+                 fs/path
+                 fs/normalize)]
+    (when-let [dir (fs/parent path)]
       (fs/create-dirs dir))
-    (write-csv out-file deps)
-    (println "Licenses written to" (str out-file))))
+    (println "Licenses written to" (str path))
+    path))
+
+(defn- prep-writer
+  "Get a io/writer for a given path, make parent directories as needed.
+  Path value `:stdout` creates a writer with the standard output."
+  [{:keys [path] :as options}]
+  (cond
+    (= path :stdout) {:writer *out* :close? false}
+    :else {:writer (-> options
+                       prep-path
+                       fs/file
+                       io/writer)
+           :close? true}))
```

Comment:
I think I would do that differently. This prints "licenses written" _before_ writing, and the `:writer` and `:close?` opts are a bit awkward. Also, we don't close the writer when there is an exception.

Maybe a better design would be to wrap the writing function:

```clojure
(defmulti ^:private write-format (fn [writer deps opts] (:format opts)))

(defmethod write-format :csv [writer deps _] ,,,)

(defn- write-deps [deps {:keys [out out-file project] :as opts}]
  (if (= :stdout out)
    (write-format *out* deps opts)
    (let [path ,,,]
      ;; ...create-dirs...
      (with-open [writer (io/writer (fs/file path))]
        (write-format writer deps opts)))
      (println "Licenses written...")))
```